lang-guide/lang-guide.md

```diff
@@ -1490,6 +1490,183 @@ It doesn't work because it sees `\P` and `\s` as escapes that are not recognized
+Much of Nu's composability comes from the special `$in` variable, which holds the current pipeline input.
+
+`$in` is particular useful when used in:
+
+- Command or external parameters
+- Filters
+- Custom command definitions or scripts that accept pipeline input
+
+#### `$in` as a command argument or as part of an expression
+
+Imagine you are creating a directory that includes tomorrow's date as part of the name. The following two commands are equivalent:
+
+```nushell
+mkdir $'((date now) + 1day | format date '%F') Report'
+```
+
+or
+
+```nushell
+date now                    # 1: today
+| $in + 1day                # 2: tomorrow
+| format date '%F'          # 3: Format as YYYY-MM-DD
+| $'($in) Report'           # 4: Format the directory name
+| mkdir $in                 # 5: Create the directory
+```
+
+While the second form may be overly verbose for this contribed example, you'll notice several advantages:
+
+- It can be composed step-by-step with a simple <kbd>↑</kbd> (up arrow) to repeat the previous command and add the next stage of the pipeline.
+- It's arguably more readable.
+- Each step can, if needed, be commented.
+- Each step in the pipeline can be [`inspect`ed for debugging](/commands/docs/inspect.html).
+
+Let's examine the contents of `$in` on each line of the above example:
+
+- On line 2, `$in` refers to the results of line 1's `date now` (a datetime value).
+- On line 4, `$in` refers to tomorrow's formatted date from line 3 and is used in an interpolated string
+- On line 5, `$in` refers to the results of line 4's interpolated string, e.g. '2024-05-14 Report'
+
+#### `$in` in filters
+
+`$in` may be modified by certain [filter commands](/commands/categories/filters.html) to provide more convenient access to the expected context. For example:
+
+```nushell
+1..10 | each {$in * 2}
+```
+
+Rather than referring to the entire range of 10 digits, the `each` filter modifies `$in` inside the closure to refer to the value of the _current iteration_.
+
+In most cases, `$in` will be the same as the closure parameter. For the `each` filter, the following example is equivalent to the one above:
+
+```nushell
+1..10 | each {|value| $value * 2}
+```
+
+However, some filters will assign an even more convenient value to `$in`. The `update` filter is one example. Using `$in` with `update` refers to the _column_ being updated, while the closure parameter refers to the entire record:
+
+```nushell
+ls | update name {|file| $file.name | str upcase}
+ls | update name {$in | str upcase}
+```
+
+With most filters, `$in` would refer to the entire `file` record (with `name`, `type`, `size`, and `modified` columns). However, with `update`, it refers specifically to the contents of the _column_ being updated, in this case `name`.
+
+#### Implicit `$in`
+
+The preceding example can be simplified even further because `$in |` is implicit at the beginning of a closure. So we can just use the more concise:
+
+```nushell
+ls | update name {str upcase}
+```
```

Comment:
I think this section is misleading because there is no hidden `$in` variable in our internal semantics. 

Instead part of the custom command semantics is that the first pipeline element of the first pipeline expression will receive the pipeline input to the custom command. (We should make sure that this is both documented in the spec and the general user documentation)

This has the distinct difference that it follows the pipelining semantics instead of the value/variable semantics. Thus it is lazy and carries the pipeline metadata (latter can be considered somewhat of an implementation detail for now)

We may implement optimizations for `$in` in the "first element of custom command"-position but from an educational perspective with the current semantics of Nushell it is much easier to teach `$in` is a special variable that you can use if needed but the pipelining behavior is expressed with `|` and the beginning of a custom command. (Teaching it that way also makes you aware of potential pitfalls, that you can not assume to receive no input in the first pipeline element of a custom command)

Reply:
Fair enough - I'll work on fixing that.  Agree that the value/variable semantics are different enough (in streaming vs. collection) to make that a bad analogy.